cluster-operator/src/main/java/io/strimzi/operator/cluster/operator/assembly/KafkaAssemblyOperator.java

```diff
@@ -1000,7 +1018,7 @@ private Future<Secret> getSecret(String namespace, String secretName)  {
          *
          * @return succeeded future if kafka pod was rolled and is ready
          */
-        Future<Void> maybeRollKafka(int replicas, Function<Pod, List<String>> podNeedsRestart) {
+        Future<Void> maybeRollKafka(int replicas, Function<Pod, RestartReasons> podNeedsRestart) {
```

Comment:
Given that we're changing this anyway, maybe it's best to just have our own interface, rather than (ab)using `Function`.

Reply:
Hi @tombentley on this, what do you imagine that interface looking like?

Reply:
Fair question. Functionally, it's just an unary method, but we can find a better name for it than `apply`. I think we also need to be careful about the naming of this interface and also `RestartReasons`, because `KafkaRoller` is used for rolling dynamic reconfiguration as well as restart. That suggests we need a more abstract verb that covers both restart and reconfigure. I'm not feeling terribly imaginative right now, but something like:

```java
/**
 * A predicate/guard on performing some rolling action (such as restart or reconfiguration) on a set of Pods.
 */
interface ActionPredicate { // or ActionGuard??
    /** 
     * Returns the reasons to perform an action on the given pod, 
     * or empty reasons if no action needs to be taken on this pod.
     * @param pod The pod to test
     * @return The reasons for performing the action.
     */
    Reasons reasonsToAct(Pod pod);
}
```

Reply:
@tombentley, yep, I get your point there. Maybe it could be renamed `PotentialRestartReasons`? As there's a lot of "maybe" throughout the call stack.

E.g, `KafkaReconciler#maybeRollKafka` leads to `KafkaRoller#rollingRestart` leads to `schedule` leads to what is now called `restartOrDynamicallyReconfigure` (it was previously `checkReconfigurability`) then to `restartIfNecessary`.